### PR TITLE
Selectors in command arguments

### PR DESCRIPTION
**Pull request by JBYoshi**
As it stands currently, Sponge commands using `GenericArguments` cannot use selectors without going out of their way. This pull request adds the ability to use selectors in `GenericArguments.player()`. It also adds `GenericArguments.entity()` and `GenericArguments.entityOrSource()`, which are made much easier due to selectors.

Requires SpongePowered/SpongeCommon#110.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        final Optional<String> nextArg = args.nextIfPresent();
+        args.setState(state);
+        
+        if (nextArg.isPresent()) {
+            if (nextArg.get().startsWith("@")) {
+                if (nextArg.get().contains("]")) {
+                    return ImmutableList.of();
+                }
+                if (nextArg.get().contains("[")) {
+                    // TODO tab complete selector arguments?
```
Comment (kenzierocks) [inline src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java]: I may be able to add this if you don't feel like it, shouldn't be too hard.

Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java]: The only real problem I had was finding out which arguments are relevant to the selector, e.g. `@a` doesn't support `type`, but `@e` and `@r` do.

Comment (kenzierocks) [inline src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java]: Probably just hardcoding, also `@p` supports it too so it may be best just to complete it for all? Not sure if we really need to separate out the different types.

Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java]: Now that I think about it, I'd agree with not separating them, so long as plugins can't add their own selectors and arguments.

Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java]: Seems like they can, so I'll need to figure out how to do value completion.

**Comment by ryantheleach** [comment]
The only issue I see is that the format of selectors has been hardcoded into the API with this commit.

@kenzierocks put in quite a lot of effort in order to abstract out the format of selectors, and provide API for parsing & building them.

Could this possibly be moved to the implementation? It would also allow for CommandBook variable targets to possibly be implemented as selectors making plugins who are selector aware be able to handle them.

**Comment by JBYoshi** [comment]
@ryantheleach Sounds good.

File: src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java [hunk 2]
```diff
@@ -1247,4 +1254,90 @@ public Text getUsage(CommandSource src) {
...
+                Optional<Entity> ret = world.getEntity(uuid);
+                if (ret.isPresent()) {
+                    return ret.get();
+                }
+            }
+            throw new IllegalArgumentException("Input value " + choice + " was not an entity");
+        }
+    
+        private Entity tryReturnSource(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            // TODO get entity from ProxySource?
```
Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: @zml2008 I thought this was added in my commands PR. What happened to it?

Comment (zml2008) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: Not all ProxySources have entities attached. You have to check if the real invoker (proxied source) is an entity.

Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: All the vanilla `ProxySource`s have an entity.
- `/execute` is run as the entity it selects. This is different from the `CommandSource` causing the command to be executed.
- `SignSource` is run as the player clicking it.

Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: @JBYoshi
How about `/su` or `/execute console`.
There is no entity connected to that. RCON and Console both have no Entity.
However we could make the method return an Optional Entity as the entity is present most of the time.

Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: How about we add a new interface, `ProxySourceTarget`, that is extended by `Entity` and `CommandSource`? That should take care of it.

Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: Now that I think about it, `ProxySource` can't always be a `LocatedSource` if the target isn't an `Entity`. I'll be making a subinterface of `ProxySource` to fix that.

File: src/main/java/org/spongepowered/api/util/command/source/LocatedProxySource.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.util.command.source;
+
+/**
+ * Located proxy sources are {@link ProxySource}s that are run at a position in
+ * the world.
+ */
+public interface LocatedProxySource extends ProxySource, LocatedSource {
```
Comment (zml2008) [inline src/main/java/org/spongepowered/api/util/command/source/LocatedProxySource.java]: or you could just check instanceof LocatedSource? might as well take advantage of multiple interface inheritance.

File: src/main/java/org/spongepowered/api/util/command/source/ProxySource.java [hunk 4]
```diff
@@ -24,20 +24,25 @@
...
      * @return The proxied source
      */
     CommandSource getOriginalSource();
+
+    /**
+     * Get the target of this source.
+     *
+     * @return The target
+     */
+    ProxySourceTarget getTarget();
```
Comment (zml2008) [inline src/main/java/org/spongepowered/api/util/command/source/ProxySource.java]: This should be covered fine by getOriginalSource

Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/source/ProxySource.java]: Actually, it's not. If PlayerA types `/execute PlayerB ~ ~ ~ /command`, `getOriginalSource()` returns PlayerA, while `getTarget()` returns PlayerB.

File: src/main/java/org/spongepowered/api/util/command/source/LocatedProxySource.java (hunk 3 above)
Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/source/LocatedProxySource.java]: Sounds good.

**Comment by Zidane** [comment]
@JBYoshi 

Can you update your PR so we can evaluate this for merge?

**Comment by JBYoshi** [comment]
Sounds good. I should have time to rebase tonight.

**Comment by JBYoshi** [comment]
@Zidane Done.

File: src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java [hunk 5]
```diff
@@ -31,11 +31,13 @@
 import com.google.common.base.Function;
 import com.google.common.base.Joiner;
 import com.google.common.base.Optional;
+import com.google.common.base.Predicates;
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: Should migrate to Java's

File: src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.util.command.args;
+
+import com.google.common.base.Optional;
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java]: Should use Java's

File: src/main/java/org/spongepowered/api/util/command/source/ProxySource.java [hunk 7]
```diff
@@ -28,14 +28,21 @@
...
  * Proxy sources are {@link CommandSource}s that are run by one source as a
- * different source.
+ * different object.
  */
-public interface ProxySource extends LocatedSource {
+public interface ProxySource extends CommandSource {
 
     /**
-     * Get the {@link CommandSource} this source is proxying. This
+     * Get the {@link CommandSource} this source is created by.
```
Comment (Phroa) [inline src/main/java/org/spongepowered/api/util/command/source/ProxySource.java]: These should be `Gets` for consistency

File: src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java (hunk 6 above)
Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/args/SelectorCommandElement.java]: I haven't updated to Java 8 yet.

File: src/main/java/org/spongepowered/api/util/command/source/ProxySource.java (hunk 7 above)
Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/source/ProxySource.java]: I didn't change that, but I'll fix it.

Comment (Phroa) [inline src/main/java/org/spongepowered/api/util/command/source/ProxySource.java]: :heart:

**Comment by JBYoshi** [comment]
@zml2008 Status?

File: src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java [hunk 8]
```diff
@@ -895,16 +901,16 @@ protected PlayerCommandElement(Text key, Game game, boolean returnSource) {
...
         protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
-            // TODO: Make player name resolution better -- support selectors, etc
             if (!args.hasNext() && this.returnSource) {
                 return tryReturnSource(source, args);
             }
 
             Object state = args.getState();
             try {
-                return super.parseValue(source, args);
+                return Iterables.filter((Iterable<Entity>)super.parseValue(source, args), Predicates.instanceOf(Player.class));
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: Couldn't you use streams or any of the new collections API?

File: src/main/java/org/spongepowered/api/util/command/source/ProxySource.java [hunk 9]
```diff
@@ -28,14 +28,21 @@
 
 /**
  * Proxy sources are {@link CommandSource}s that are run by one source as a
- * different source.
+ * different object.
  */
-public interface ProxySource extends LocatedSource {
+public interface ProxySource extends CommandSource {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/util/command/source/ProxySource.java]: Why the change? (Out of curiosity).

File: src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java (hunk 8 above)
Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: I wrote this before the java 8 update. I'll change it later.

File: src/main/java/org/spongepowered/api/util/command/source/ProxySource.java (hunk 9 above)
Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/source/ProxySource.java]: It's for those cases where the target is not in a location. There was a conversation in an "outdated diff" about that.

File: src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java (hunk 8 above)
Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java]: Changed this to lambdas, but I couldn't use streams because I might be given an `Iterable` (which doesn't support streams) as opposed to a `Collection`.

**Comment by JBYoshi** [comment]
Status?

**Comment by gabizou** [comment]
Merged in https://github.com/SpongePowered/SpongeAPI/commit/221b0b7b5166aaef1f56cdad2a04e01820afe044

**Comment by JBYoshi** [comment]
Status?
